permission-element.bs

```diff
@@ -19,7 +18,408 @@ Abstract: A `<permission>` HTML element to request browser permissions in-page.
+<tr><th>type_count
+<td>permanent
+<td>When another <{permission}> for the same [=powerful feature|permission type=] is used on the same page.
+<tr><th>illegal_subframe
+<td>permanent
+<td>When the <{permission}> is used inside a <{fencedframe}>.
+<tr><th>covered
+<td>temporary
+<td>When the <{permission}> is partially occluded, or not visible at all.
+<tr><th>recently_moved
+<td>expiring
+<td>When the <{permission}> is being moved on the screen.
+<tr><th>recently_created
+<td>expiring
+<td>When <{permission}> is attached to the document.
+</table>
+</div>
+
+<div algorithm>
+To <dfn for="HTMLPermissionElement">add a blocker</dfn> with a
+{{PermissionElementInvalidReason}} |reason| and an optional flag |expires|:
+
+1. [=Assert=]: |reason| is not `""`.
+    (The empty string in {{PermissionElementInvalidReason}} signals no blocker
+    is present. Why would you add a non-blocking blockern empty string?)
+1. Let |timestamp| be None.
+1. If |expires|, then let |timestamp| be [=current high resolution time=]
+    plus the [=blocker delay=].
+1. [=list/Append=] an entry to the internal {{[[BlockerList]]}} with |reason|
+    and |timestamp|.
+
+</div>
+
+<div>
+The <dfn for="HTMLPermissionElement">blocker delay</dfn> is 500ms.
+</div>
+
+<div algorithm>
+To <dfn for="HTMLPermissionElement">add an expiring blocker</dfn> with a
+{{PermissionElementInvalidReason}} |reason|:
+
+1. [=Assert=]: |reason| is listed as "expiring" in the [=blocker reason table=].
+1. [=Add a blocker=] with |reason| and true.
+
+</div>
+
+<div algorithm>
+To <dfn for="HTMLPermissionElement">add a temporary blocker</dfn> with a
+{{PermissionElementInvalidReason}} |reason|:
+
+1. [=Assert=]: |reason| is listed as "temporary" in the [=blocker reason table=].
+1. [=Add a blocker=] with |reason| and false.
+
+</div>
+
+<div algorithm>
+To <dfn for="HTMLPermissionElement">add a permament blocker</dfn> with a
+{{PermissionElementInvalidReason}} |reason|:
+
+1. [=Assert=]: |reason| is listed as "permanent" in the [=blocker reason table=].
+1. [=Add a blocker=] with |reason| and false.
+
+</div>
+
+<div algorithm>
+To <dfn for="HTMLPermissionElement">remove blockers</dfn> with
+{{PermissionElementInvalidReason}} |reason| from an |element|:
+
+1. [=Assert=]: |reason| is listed as "temporary" in the [=blocker reason table=].
+1. [=list/iterate|For each=] |entry| in |element|'s {{[[BlockerList]]}}:
+    1. If |entry|'s reason [=string/is|equals=] |reason|, then [=list/remove=]
+        |entry| from |element|'s {{[[BlockerList]]}}.
+
+</div>
+
+<div algorithm>
+To determine a {{HTMLPermissionElement}} |element|'s
+<dfn for="HTMLPermissionElement">blocker</dfn>:
+
+1. [=list/iterate|For each=] |entry| in |element|'s {{[[BlockerList]]}}:
+    1. If |entry| is [=HTMLPermissionElement/valid=], then return |entry|.
+1. Return nothing.
+
+</div>
+
+<div algorithm>
+An {{HTMLPermissionElement}}'s [=blocker=] list's |entry| is
+<dfn for="HTMLPermissionElement">valid</dfn> if:
```

Comment:
We use the term "valid" here for blockers to signify that the blocker is making the element "invalid" (aka the blocker is currently in effect), but that terminology seems confusing. Should we switch to a different term like "active" or "in effect" or "expired" (and flipping the logic where blocker#valid is used to !blocker#expired).

Reply:
Hmm. I now changed this to "blocking", so a blocker is blocking (or not).